dev-tools/harness.js

```diff
@@ -0,0 +1,241 @@
+			`>player p2 ${JSON.stringify(p2spec)}`);
+
+		let chunk;
+		while ((chunk = await streams.omniscient.read())) {
+			if (this.output) console.log(chunk);
+		}
+	}
+
+	// Same as PRNG#generatedSeed, only deterministic.
+	// NOTE: advances this.prng's seed by 4.
+	newSeed() {
+		return [
+			Math.floor(this.prng.next() * 0x10000),
+			Math.floor(this.prng.next() * 0x10000),
+			Math.floor(this.prng.next() * 0x10000),
+			Math.floor(this.prng.next() * 0x10000),
+		];
+	}
+
+	getPlayerSpec(name, options) {
+		if (options.team) return {name, team: options.team};
+		return {name, seed: this.newSeed()};
+	}
+}
+
+class RawBattleStream extends BattleStreams.BattleStream {
+	constructor(input) {
+		super();
+		this.input = !!input;
+		this.rawInputLog = [];
+	}
+
+	_write(message) {
+		if (this.input) console.log(message);
+		this.rawInputLog.push(message);
+		super._write(message);
+	}
+}
+
+class MultiRunner {
+	constructor(options) {
+		this.options = Object.assign({}, options);
+
+		this.totalGames = options.totalGames;
+
+		this.prng = (options.prng && !Array.isArray(options.prng)) ?
+			options.prng : new PRNG(options.prng);
+		this.options.prng = this.prng;
+
+		this.format = options.format;
+		this.cycle = !!options.cycle;
+		this.all = !!options.all;
+
+		this.async = !!options.async;
+
+		this.formatIndex = 0;
+		this.numGames = 0;
+	}
+
+	async run() {
+		let games = [];
+		let format, lastFormat;
+		let failures = 0;
+		while ((format = this.getNextFormat())) {
+			if (this.all && lastFormat && format !== lastFormat) {
+				if (this.async) await Promise.all(games);
+				games = [];
+			}
+
+			const seed = this.prng.seed;
+			const game = new Runner(Object.assign({format}, this.options)).run().catch(err => {
+				failures++;
+				console.error(
+					`Run \`node dev-tools/harness 1 --format=${format} --seed=${seed.join()}\` ` +
+					`to debug (optionally with \`--output\` and/or \`--input\` for more info):\n`, err);
+			});
+
+			if (!this.async) await game;
+			games.push(game);
+			lastFormat = format;
+		}
+
+		if (this.async) await Promise.all(games);
+		return failures;
+	}
+
+	getNextFormat() {
+		if (this.formatIndex > FORMATS.length) return false;
+
+		if (this.numGames++ < this.totalGames) {
+			if (this.format) {
+				return this.format;
+			} else if (this.all) {
+				return FORMATS[this.formatIndex];
+			} else if (this.cycle) {
+				const format = FORMATS[this.formatIndex];
+				this.formatIndex = (this.formatIndex + 1) % FORMATS.length;
+				return format;
+			} else {
+				return this.prng.sample(FORMATS);
+			}
+		} else if (this.all) {
+			this.numGames = 1;
+			this.formatIndex++;
+			return FORMATS[this.formatIndex];
+		}
+
+		return false;
+	}
+}
+
+// Tracks whether some promises threw errors that weren't caught so we can log
+// and exit with a non-zero status to fail any tests. This "shouldn't happen"
+// because we're "great at propagating promises (TM)", but better safe than sorry.
+const RejectionTracker = {
```

Comment:
Hm, you recently switched this from `new class {` to `{`. Want to talk about your reasoning?

I'm still undecided which is better. I'm leaning towards `new class {` because it feels more natural in TypeScript.

Reply:
There's also `class` with literally everything `static`. Which is also an option I'm undecided about.

Reply:
https://github.com/Zarel/Pokemon-Showdown/pull/5370#discussion_r271503441 is my reasoning.

Reply:
I'm also happy to follow up and bikeshed this to look however you want, but I have several PRs blocked on this landing so it would be awesome if we could merge first and argue over the insignificant bits later :)